task01/src/Vector.cpp

```diff
@@ -0,0 +1,318 @@
+        delete[] m_array;
+        m_array = new_array;
+        m_capacity = new_capacity;
+    }
+
+    // Add the new element to the end of the array and update the size
+    m_array[m_size] = value;
+    m_size++;
+}
+
+//--------------------------------------------------------------------------------
+
+void Vector::resize(std::size_t count) {
+	resize(count, 0);
+	
+}
+
+//--------------------------------------------------------------------------------
+
+void Vector::resize(std::size_t count, const int& value) {
+    if (count <= m_capacity) {
+        if (count < m_size) {
+            // Truncate the vector by updating its size
+            m_size = count;
+        } else if (count > m_size) {
+            // Append the specified value to the vector for the remaining elements
+            for (std::size_t i = m_size; i < count; ++i) {
+                m_array[i] = value;
+            }
+            m_size = count;
+        }
+    } else {
+        // Allocate a new buffer with appropriate size
+        std::size_t new_capacity = m_capacity * 2;
+        while (new_capacity < count) {
+            new_capacity *= 2;
+        }
+
+        // Create new array and copy elements from old array to new array
+        int* new_array = new int[new_capacity];
+        for (std::size_t i = 0; i < m_size; ++i) {
+            new_array[i] = m_array[i];
+        }
+        // Append the specified value to the vector for the remaining elements
+        for (std::size_t i = m_size; i < count; ++i) {
+            new_array[i] = value;
+        }
+
+        // Deallocate the old buffer and update the vector's fields
+        delete[] m_array;
+        m_array = new_array;
+        m_capacity = new_capacity;
+        m_size = count;
+    }
+}
+
+//--------------------------------------------------------------------------------
+/* data() function whitch returns a pointer of first cell*/
+int* Vector::data() const {
+	if (m_size != 0) {
+		return &m_array[0];
+	}
+}
+
+//--------------------------------------------------------------------------------
+void Vector::clear() {
+	m_size = 0;
+	m_array = 0;
+}
+
+//--------------------------------------------------------------------------------
+
+void Vector::pop_back() {
+    if (m_size > 0) {
+        --m_size;
+    }
+}
+
+//--------------------------------------------------------------------------------
+/*check if left = right */
+bool operator==(const Vector& lhs, const Vector& rhs) {
+    if (lhs.getSize() != rhs.getSize()) {
+        return false;
+    }
+
+    for (std::size_t i = 0; i < lhs.getSize(); ++i) {
+        if (lhs[i] != rhs[i]) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+//--------------------------------------------------------------------------------
+/*this function inserts a value in a giving possision.*/
+int* Vector::insert(iterator pos, const int& val) {
+    // calculate the index of the position iterator
+    int index = pos - begin();
```

Comment:
What if `pos` is not in the array's range? In that case `index`'s value is undefined